File: src/reassembler.cc
#include "reassembler.hh"

using namespace std;

Reassembler::Segment::Segment( uint64_t first, std::string data, bool eof )
  : first_( first ), last_( first + data.length() ), data_( data ), eof_( eof )
{}

Reassembler::Segment::Segment( uint64_t first, uint64_t last, std::string data, bool eof )
  : first_( first ), last_( last ), data_( data ), eof_( eof )
{}

Reassembler::Reassembler() : segs_() {}

void Reassembler::insert( uint64_t first_index, string data, bool is_last_substring, Writer& output )
{
  uint64_t min_index = output.bytes_pushed();
  uint64_t max_index = min_index + output.available_capacity();
  uint64_t last_index = first_index + data.length();

  if ( first_index > max_index || last_index < min_index )
    return;

  if ( last_index > max_index ) {
    data = data.substr( 0, max_index - first_index );
    last_index = max_index;
  }
  if ( first_index < min_index ) {
    data = data.substr( min_index - first_index );
    first_index = min_index;
  }

  list<Segment> new_segs;
  bool segment_pushed = false;
  for ( const Segment& s : segs_ ) {
    if ( segment_pushed || s.last_ < first_index ) {
      new_segs.push_back( s );
      continue;
    }
    if ( s.first_ > last_index ) {
      new_segs.push_back( Segment( first_index, last_index, data, is_last_substring ) );
      segment_pushed = true;
      new_segs.push_back( s );
      continue;
    }
    if ( s.first_ < first_index ) {
      data = s.data_.substr( 0, first_index - s.first_ ) + data;
      first_index = s.first_;
    }
    if ( is_last_substring )
      break;
    if ( s.last_ >= last_index ) {
      data += s.data_.substr( last_index - s.first_ );
      last_index = s.last_;
      is_last_substring = s.eof_;
    }
  }
  if ( !segment_pushed )
    new_segs.push_back( Segment( first_index, last_index, data, is_last_substring ) );
  swap( new_segs, segs_ );

  const Segment& front = segs_.front();
  if ( front.first_ == min_index ) {
    output.push( front.data_ );
    if ( front.eof_ )
      output.close();
    segs_.pop_front();
  }
}

uint64_t Reassembler::bytes_pending() const
{
  uint64_t bytes = 0;
  for ( const Segment& s : segs_ ) {
    bytes += s.data_.length();
  }
  return bytes;
}

File: pytcp/tcp/reassembler.py
from collections import deque
from dataclasses import dataclass
from typing import Optional

from .stream import Stream


@dataclass
class _Segment:
    first: int
    last: int
    data: bytes
    eof: bool

    def narrow(self, min_idx: int, max_idx: int) -> Optional['_Segment']:
        if self.first > max_idx or self.last < min_idx:
            return None
        seg = _Segment(self.first, self.last, self.data, self.eof)
        if seg.last > max_idx:
            seg.data = seg.data[:max_idx - seg.first]
            seg.last = max_idx
        if seg.first < min_idx:
            seg.data = seg.data[min_idx - seg.first:]
            seg.first = min_idx
        return seg


class Reassembler(Stream):
    __segs: deque[_Segment]

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.__segs = deque()

    def insert(self, idx: int, data: bytes, eof: bool = False):
        self.insert_segment(_Segment(idx, idx + len(data), data, eof))

    def insert_segment(self, seg: _Segment):
        min_idx = self.bytes_pushed()
        max_idx = min_idx + self.cap()

        _seg: Optional[_Segment] = seg.narrow(min_idx, max_idx)
        if _seg is None:
            return
        seg = _seg

        segs: deque[_Segment] = deque()
        seg_appended = False

        for s in self.__segs:
            if seg_appended or s.last < seg.first:
                segs.append(s)
                continue
            if s.first > seg.last:
                segs.append(seg)
                seg_appended = True
                segs.append(s)
                continue
            if s.first < seg.first:
                seg.data = s.data[:seg.first - s.first] + seg.data
                seg.first = s.first
            if seg.eof:
                break
            if s.last >= seg.last:
                seg.data += s.data[seg.last - s.first:]
                seg.last = s.last
                seg.eof = s.eof

        if not seg_appended:
            segs.append(seg)

        self.__segs = segs

        if seg.first == min_idx:
            self.__segs.popleft()
            if len(seg.data) != 0:
                self.push(seg.data)
            if seg.eof:
                self.close()

    def bytes_pending(self) -> int:
        return sum(len(seg.data) for seg in self.__segs)
